### PR TITLE
feat: session.stuck event with lastTool, lastToolError, tokenDelta (fixes #1585)

### LINKED ISSUE #1585: feat: session.stuck event with lastTool, lastToolError, tokenDelta
## Summary

Sprint 38 had a repair session stall for 2+ minutes retrying a tool (`Monitor`) not in its toolset — the orchestrator noticed only by accident. Emit `session.stuck` after N seconds without token progress, carrying `lastTool`, `lastToolError`, and `tokenDelta` so the orchestrator can interrupt and redirect.

## Surface area

| File | Change |
|---|---|
| `packages/daemon/src/claude-session/session-state.ts` | Track per-session `lastToolCall: {name: string, errorMessage?: string, at: number}` and `lastTokenCount: number`. Update on each `tool_use` event and `message` event. |
| `packages/daemon/src/claude-session/ws-server.ts` | Per-session idle watchdog: timer reset on each token/tool progress event. After 90s of no progress, emit `session.stuck` to the bus. Re-arm at 90s + 180s + 300s (subsequent stuck events escalate). |
| `packages/core/src/monitor-event.ts` | Add `SESSION_STUCK = \"session.stuck\"` constant. |

## Stuck-detection heuristic

Track two signals per session:
- `lastProgressAt` = most recent timestamp of a tool-use, tool-result, or assistant-message event
- `tokenDelta` = tokens since last stuck check (should be >0 on an active session)

Fire `session.stuck` when `now - lastProgressAt > 90s` AND session is not awaiting a permission request. Escalate tiers at 90s / 180s / 300s — each tier includes the elapsed duration in the payload.

## Payload shape

```json
{"event":"session.stuck","category":"session",
 "sessionId":"fcfbc19d","workItemId":"#1441",
 "sinceMs":180000,
 "lastTool":"Monitor","lastToolError":"No matching rule for tool: Monitor",
 "tokenDelta":12,
 "tier":2}
```

`tier` = 1 (90s), 2 (180s), 3 (300s+). Past tier 3, re-emit every 300s.

## Acceptance criteria

- [ ] `session.stuck` fires exactly once per tier per stall (no dup emissions on flapping)
- [ ] Stall detector pauses while session is awaiting a permission request (that's not \"stuck\", it's blocked on user)
- [ ] `lastTool` captures the *name* of the last tool call attempted; `lastToolError` captures the error message only if the tool call itself failed (not if tool use succeeded but assistant is \"thinking\")
- [ ] `tokenDelta` is the count since last emission (or since session start for tier 1)
- [ ] Session resumes: progress event clears all tiers; next stall starts fresh at tier 1
- [ ] Watchdog timer cleanly disposed on session end / daemon shutdown
- [ ] Unit test: progress resets timer
- [ ] Unit test: permission request suspends detector
- [ ] Unit test: tier escalation at 90s, 180s, 300s
- [ ] Integration test: spawn a session with a blocked tool call, observe tier 1 event after 90s

## Dependencies

- Depends on #1562 (session events → main-thread bus) — stuck events must reach the IPC stream
- Complements P6-daemon-lifecycle (both are \"supervisory\" events)

## Scrutiny / model

**Scrutiny: medium** — timer hygiene + false-positive avoidance (permission requests, long-running tools like `Bash` on a real build). Get the heuristic wrong and you spam the orchestrator.
**Model: sonnet** — well-scoped once the heuristic is specified.

### PR DESCRIPTION
**Pull request by theshadow27**
## Summary
- Emit `session.stuck` monitor event at escalating tiers (90s / 180s / 300s, then repeat) when a session stalls with no token/tool progress
- `StuckDetector` class with explicit `dispose()` owns all timer handles — disposed on every exit path: `session:result`, `session:error`, `session:ended`, `session:disconnected`, `session:cleared`, and `terminateSession`
- Track `lastToolCall` (name + errorMessage from permission denials) on `SessionState` for stuck event payload
- Watchdog pauses while session is `waiting_permission` (not stuck, just blocked on user)

## Test plan
- [x] 14 unit tests for `StuckDetector`: tier escalation, progress reset, permission suspension, dispose/leak prevention, tokenDelta tracking, flapping
- [x] 6 `lastToolCall` unit tests in session-state.spec.ts: extraction from tool_use blocks, error from permission denial
- [x] 4 integration tests in ws-server.spec.ts: stuck event fires after threshold, no timer leak after session:result, no timer leak after bye (session:ended), workItemId propagation
- [x] typecheck, lint, full test suite (5938 tests pass)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by theshadow27** [comment]
## Adversarial Review (Delta)

### Changes Since Last Review

| # | Previous Issue | Status | Notes |
|---|----------------|--------|-------|
| 1 | 🟡 `lastToolError` only captures permission denials, not MCP execution errors | ✅ Addressed in `c960322b` | MCP execution errors are architecturally invisible at the NDJSON layer (tool results flow through MCP transport in server-pool, not through the session WS). Added a comment at `session-state.ts:151` documenting the limitation and pointing to the follow-up. Behavior is correct and documented; wiring server-pool errors is a separate tracked concern. |
| 2 | 🟡 `resolveEventWaiters` for stuck events drops all diagnostic fields from waiter payload | ✅ Fixed in `c960322b` | Added `tier`, `sinceMs`, `tokenDelta`, `lastTool`, `lastToolError` to `SessionWaitEvent` interface. `handleStuckEvent` now populates all five fields in the `resolveEventWaiters` call. New test `waitForEvent resolves with stuck diagnostic fields` verifies the waiter payload. |
| 3 | 🔵 No constructor validation for empty `thresholdsMs` | ✅ Fixed in `c960322b` | Constructor now throws `"thresholdsMs must be non-empty"` if passed an empty array. New test `constructor throws on empty thresholdsMs` covers it. |
| 4 | 🔵 `evaluate()` missing `disposed` re-check before `scheduleNext()` | ✅ Fixed in `c960322b` | Added `if (this.disposed) return` before the final `scheduleNext()` call in `evaluate()`. |
| 5 | 🔵 `Date.now()` susceptible to NTP/sleep clock jumps | ✅ Fixed in `c960322b` | Switched `lastProgressAt` and `elapsed` to `performance.now()` (monotonic). |
| 6 | 🔵 Negative tests use `Bun.sleep` fixed delays | ⏳ Not addressed | These are unavoidable for negative assertions (cannot `pollUntil` for absence of events). Timing margins are 4× the threshold so flake risk is low in practice. Not worth a redesign. |

### New Issues
No new issues introduced.

### Suggested Automation
No new automation candidates beyond what was noted in the original review.

### Updated Verdict
✅ **Approve** — both 🟡 issues resolved (waiter payload now carries full stuck diagnostic; MCP execution error gap documented), all three 🔵 suggestions applied. The remaining `Bun.sleep` item is a policy note, not a reliability concern.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1761 — feat: session.stuck event with lastTool, lastToolError, tokenDelta (fixes #1585)
**Linked issue:** #1585

### Implementation Evidence
- `packages/core/src/monitor-event.ts`: `SESSION_STUCK = "session.stuck"` constant added; `FORMATTERS[SESSION_STUCK]` formatter renders tier, sinceMs, lastTool, lastToolError
- `packages/daemon/src/claude-session/stuck-detector.ts`: New `StuckDetector` class with explicit `dispose()`, `recordProgress()`, tier escalation (90s/180s/300s/repeat), permission suspension, `performance.now()` monotonic clock, and empty-thresholds constructor guard
- `packages/daemon/src/claude-session/session-state.ts:84`: `lastToolCall` field added; `extractLastToolCall()` / `extractLastToolCallRaw()` extract last `tool_use` block from assistant messages; `respondToPermission` sets `errorMessage` on denial
- `packages/daemon/src/claude-session/ws-server.ts`: `stuckDetector: StuckDetector | null` on `WsSession`; `ensureStuckDetector()`, `recordSessionProgress()`, `disposeStuckDetector()`, `handleStuckEvent()` methods; `disposeStuckDetector` called on `session:result`, `session:error`, `session:cleared`, `session:disconnected`, and `terminateSession`; `recordSessionProgress` called on `session:init`, `session:response`, `session:permission_request`, `queuePrompt`, `respondToPermission`; `SessionWaitEvent` extended with `tier`, `sinceMs`, `tokenDelta`, `lastTool`, `lastToolError`

### Test Coverage
- `packages/daemon/src/claude-session/stuck-detector.spec.ts`: 14 unit tests — tier escalation (1→2→3), tier-3 repeat, progress reset, permission suspension, non-active state stops detection, tokenDelta tracking, lastTool/lastToolError from snapshot, dispose idempotency, timer leak prevention, recordProgress-after-dispose no-op, flapping, constructor validation
- `packages/daemon/src/claude-session/session-state.spec.ts`: 6 `lastToolCall` tests — starts null, extracts last tool_use block, multiple tool_use blocks (takes last), no tool_use unchanged, permission denial sets errorMessage, permission approval does not set errorMessage
- `packages/daemon/src/claude-session/ws-server.spec.ts`: 5 integration tests — `session.stuck` fires after stall threshold, no timer leak after `session:result`, no timer leak after bye (`session:ended`), `workItemId` propagation, `waitForEvent` resolves with full stuck diagnostic fields

### Test Results
CI passed: check ✓ (4m3s), coverage ✓ (5m53s), build ✓ (11s)
[CI run](https://github.com/theshadow27/mcp-cli/actions/runs/24910368146)

### Copilot Inline Threads — BLOCKED (2 unreplied)

**Thread #3140106079** — `ws-server.ts` line 1641:
> `disposeStuckDetector()` in the `session:disconnected` event handler likely won't run for the common WebSocket-close path: `disconnectSessionWs()` transitions the state to "disconnected" and calls `onSessionEvent`, but it does not route the resulting event back through `handleSessionEvent()`. As a result...

**Thread #3140106096** — `stuck-detector.ts` line 135:
> `nextDelayMs()` assumes `config.thresholdsMs` is non-empty and strictly increasing. Because `stuckConfig` is injectable, an empty array (or non-monotonic values) will produce `undefined`/negative delays and can cause immediate tight-loop emissions. Consider validating the config in the constructor (...

Both threads have no reply from `theshadow27`. Per QA gate rules, `qa:pass` is blocked until each thread has an owner reply (fix citation or explicit dismissal).

### CI Status
✅ All 3 checks pass (build, check, coverage)

### Remaining Work (non-blocking)
- MCP execution errors (tool throws, non-zero exit) not surfaced via `lastToolError` — architecturally invisible at NDJSON layer, documented in session-state.ts:151, tracked as follow-up per issue description
- Negative tests use `Bun.sleep` fixed delays (per repair session: 4× threshold margins, considered acceptable)

### Outcome
**qa:fail** — 2 unreplied Copilot inline threads (IDs 3140106079, 3140106096). Action for implementation: reply to each thread on the PR (with a fix commit citation or explicit dismissal), then request re-QA.

File: packages/daemon/src/claude-session/ws-server.ts [hunk 1]
```diff
@@ -1612,6 +1631,7 @@ export class ClaudeWsServer {
         }
         break;
       case "session:disconnected":
+        this.disposeStuckDetector(session);
         try {
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/claude-session/ws-server.ts:1652]: Fixed in 050ad3b7. `disconnectSessionWs` was calling `onSessionEvent` but not `handleSessionEvent`, so the `session:disconnected` branch in `handleSessionEvent` never ran on WS close. Added a direct `disposeStuckDetector(session)` call at the top of `disconnectSessionWs`, mirroring the pattern already used in `terminateSession`. Timer is now cleared immediately on every WS-close path.

File: packages/daemon/src/claude-session/stuck-detector.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  private tierForEmission(emissionCount: number): number {
+    return Math.min(emissionCount + 1, this.config.thresholdsMs.length);
+  }
+
+  private nextDelayMs(emissionCount: number): number {
+    const { thresholdsMs, repeatMs } = this.config;
+    if (emissionCount === 0) return thresholdsMs[0];
+    if (emissionCount < thresholdsMs.length) {
+      return thresholdsMs[emissionCount] - thresholdsMs[emissionCount - 1];
+    }
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/claude-session/stuck-detector.ts:152]: Fixed in 050ad3b7. Added two constructor guards: (1) strictly ascending check — iterates thresholdsMs and throws if any element is ≤ its predecessor, preventing negative setTimeout delays; (2) repeatMs > 0 check. Both throw with descriptive messages. Tests added for all three invalid-config cases (empty, non-ascending, non-positive repeatMs).

**Comment by theshadow27** [comment]
## Repair delta — 050ad3b7

Both Copilot threads addressed:

**Thread #3140106079 — timer leak on WS disconnect**

`disposeStuckDetector` was only called from `handleSessionEvent` (the `session:disconnected` case), but `disconnectSessionWs` — the path taken on every WS close — calls `onSessionEvent` only, never `handleSessionEvent`. Timer could stay live until the next tick.

Fix: added `this.disposeStuckDetector(session)` at the top of `disconnectSessionWs`, mirroring the pattern in `terminateSession`.

**Thread #3140106096 — non-monotonic config causes tight-loop emissions**

Constructor only validated non-empty `thresholdsMs` but not ascending order or `repeatMs > 0`. A non-monotonic array produces negative `setTimeout` delays (treated as 0 by the runtime), flooding the event loop.

Fix: added constructor guards for strictly ascending thresholds and positive `repeatMs`, both with descriptive error messages. Tests added for all three invalid-config cases.

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1761 — feat: session.stuck event with lastTool, lastToolError, tokenDelta (fixes #1585)
**Linked issue:** #1585

### Implementation Evidence
- `packages/core/src/monitor-event.ts`: `SESSION_STUCK = "session.stuck"` constant added (line 33)
- `packages/daemon/src/claude-session/stuck-detector.ts`: new `StuckDetector` class with `recordProgress()`, `dispose()`, tier escalation at configurable thresholds, permission suspension, token delta tracking
- `packages/daemon/src/claude-session/session-state.ts`: `lastToolCall: {name, errorMessage?, at}` field; `extractLastToolCall()` on assistant message events; `errorMessage` set on permission denial
- `packages/daemon/src/claude-session/ws-server.ts`: `stuckDetector: StuckDetector | null` per `WsSession`; `ensureStuckDetector()`, `recordSessionProgress()`, `disposeStuckDetector()`, `handleStuckEvent()` methods; dispose on `session:result`, `session:error`, `session:cleared`, `session:disconnected`, `terminateSession`; `SESSION_STUCK` event forwarded to `onMonitorEvent` bus; `SessionWaitEvent` extended with `tier`, `sinceMs`, `tokenDelta`, `lastTool`, `lastToolError`

### Test Coverage
- `stuck-detector.spec.ts`: 14 unit tests — tier escalation 1→2→3→repeat, progress reset, permission suspension, dispose/leak prevention, tokenDelta, flapping, constructor validation (empty thresholds, non-ascending thresholds, non-positive repeatMs)
- `session-state.spec.ts`: 6 `lastToolCall` tests — extraction from tool_use blocks, last tool selected from multiple blocks, no tool leaves field null, permission denial sets errorMessage, successful permission doesn't set errorMessage
- `ws-server.spec.ts`: 4 integration tests — stuck event fires after threshold, no timer leak after `session:result`, no timer leak after `bye`, workItemId propagation, `waitForEvent` resolves with stuck diagnostic fields

### Test Results
- Targeted tests (stuck-detector + session-state + ws-server): **78 + 183 = 261 tests, all pass**
- Full suite from PR worktree: **5941 pass, 1 fail** — the failure is `containment.spec.ts` "symlink traversal" test that is environment-sensitive when run from `/tmp` (pre-existing; tracked in #1687/#1689/#1743/#1770; same failure occurs on `main` when run from `/tmp`)

### CI Status
All 3 required checks pass: `build` ✓, `check` ✓ (4m5s), `coverage` ✓ (6m0s)

### Copilot Gate — BLOCKED

4 unreplied Copilot inline threads (programmatic gate, no exceptions):

| Thread ID | File | Line | Issue |
|-----------|------|------|-------|
| 3140199686 | `stuck-detector.ts` | 58 | `thresholdsMs` values ≤ 0 not validated — strictly-ascending check passes `[0, 100]`; zero/negative threshold would fire immediately and spam |
| 3140199719 | `session-state.ts` | 159 | `lastToolCall.errorMessage` set unconditionally on any denial, but `lastToolCall` may point to a *different* tool than the one being denied — incorrect `lastToolError` in stuck payloads |
| 3140199742 | `ws-server.ts` | 1876 | `handleStuckEvent()` calls `resolveEventWaiters()` without try/catch; exception in `setTimeout` callback = unhandled rejection = potential daemon crash |
| 3140199756 | `ws-server.ts` | 1842 | Watchdog doesn't reset on `tool_progress`/`stream_event` (in `IGNORED_TYPES`); long-running streaming tools can trigger false-positive stuck events |

### Outcome
**qa:fail** — 4 unreplied Copilot threads block `qa:pass` per Step 5b gate. CI is green; implementation is substantively correct. Action for implementation: address each of the 4 Copilot threads (or reply with explicit dismissal + filed follow-up issue for each deferred nit).

File: packages/daemon/src/claude-session/stuck-detector.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    for (let i = 1; i < config.thresholdsMs.length; i++) {
+      if (config.thresholdsMs[i] <= config.thresholdsMs[i - 1]) {
+        throw new Error(
+          `StuckDetectorConfig.thresholdsMs must be strictly ascending (index ${i}: ${config.thresholdsMs[i]} <= ${config.thresholdsMs[i - 1]})`,
+        );
+      }
+    }
+    if (config.repeatMs <= 0) {
+      throw new Error(`StuckDetectorConfig.repeatMs must be positive (got ${config.repeatMs})`);
+    }
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/claude-session/stuck-detector.ts:65]: Fixed in 7b1f149b. Added a positive-finite guard that iterates all thresholdsMs values before the ascending check, throwing if any is <= 0 or non-finite. Also added test cases for 0, negative, and Number.POSITIVE_INFINITY entries.

File: packages/daemon/src/claude-session/session-state.ts [hunk 4]
```diff
@@ -147,7 +148,15 @@ export class SessionState {
...
-    return permissionDeny(requestId, message ?? "Denied by session controller");
+    const denyMessage = message ?? "Denied by session controller";
+    if (this.lastToolCall) {
+      // errorMessage only captures permission denials. MCP tool execution errors
+      // (e.g. tool throws, non-zero exit) are not visible at the NDJSON layer —
+      // they flow through the MCP transport in server-pool and would need separate
+      // wiring to surface here (tracked in #1585 follow-up).
+      this.lastToolCall.errorMessage = denyMessage;
+    }
+    return permissionDeny(requestId, denyMessage);
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/claude-session/session-state.ts:159]: Fixed in 7b1f149b. Changed the guard to `if (this.lastToolCall && request.tool_name === this.lastToolCall.name)` so errorMessage is only attached when the denied request's tool_name matches the tool that set lastToolCall. Added a test with mismatched tool names to confirm errorMessage stays undefined.

File: packages/daemon/src/claude-session/ws-server.ts [hunk 5]
```diff
@@ -1787,6 +1817,66 @@ export class ClaudeWsServer {
...
+        event: SESSION_STUCK,
+        category: "session",
+        sessionId,
+        workItemId,
+        tier: event.tier,
+        sinceMs: event.sinceMs,
+        tokenDelta: event.tokenDelta,
+        lastTool: event.lastTool,
+        lastToolError: event.lastToolError,
+      });
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/claude-session/ws-server.ts]: Fixed in 7b1f149b. Wrapped the entire handleStuckEvent body in a try/catch that logs via this.logger.error on failure, matching the pattern used in all other session-event branches. Any throw from resolveEventWaiters or onMonitorEvent is now swallowed and logged rather than propagating out of the setTimeout callback.

File: packages/daemon/src/claude-session/ws-server.ts [hunk 6]
```diff
@@ -1787,6 +1817,66 @@ export class ClaudeWsServer {
...
+        (event) => this.handleStuckEvent(sessionId, session, event),
+      );
+    }
+    return session.stuckDetector;
+  }
+
+  private recordSessionProgress(sessionId: string, session: WsSession): void {
+    const detector = this.ensureStuckDetector(sessionId, session);
+    detector.recordProgress(session.state.tokens);
+  }
```
[the remark this answers is not in the document]
Comment (theshadow27) [reply packages/daemon/src/claude-session/ws-server.ts:1849]: Fixed in 7b1f149b. handleMessage now calls recordSessionProgress when msg.type === "tool_progress" || msg.type === "stream_event" before the event-loop body. These types remain in IGNORED_TYPES so the state machine is unaffected; only the stuck detector's timer is reset. Prevents false stuck fires during long-running tools or streaming responses.

**Comment by theshadow27** [comment]
## Repair delta — commit 7b1f149b

Four issues from the latest QA round, all fixed:

| Thread | Issue | Fix |
|--------|-------|-----|
| 3140199686 | Zero-threshold validation gap in `StuckDetector` constructor | Added per-element positive-finite guard before the ascending check; throws if any `thresholdsMs[i] <= 0` or non-finite. Tests added. |
| 3140199719 | Wrong-tool error attribution on permission denial | `respondToPermission` now only sets `lastToolCall.errorMessage` when `request.tool_name === this.lastToolCall.name`; mismatched tool names leave `errorMessage` untouched. Test added. |
| 3140199742 | Missing try/catch in `handleStuckEvent` timer callback | Wrapped full body of `handleStuckEvent` in try/catch → logs via `this.logger.error`, matches every other event-handling branch; can no longer take down the daemon. |
| 3140199756 | False positives from streaming `tool_progress`/`stream_event` | `handleMessage` now calls `recordSessionProgress` when either type arrives, resetting the watchdog timer during active tool execution/streaming. Types stay in `IGNORED_TYPES`; state machine unchanged. |

All pre-commit checks pass (typecheck, lint, tests, coverage).

**Comment by theshadow27** [comment]
## QA Verification

**PR:** #1761 — feat: session.stuck event with lastTool, lastToolError, tokenDelta (fixes #1585)
**Linked issue:** #1585
**Round:** 3

### Implementation Evidence
- `packages/core/src/monitor-event.ts:33`: `SESSION_STUCK = "session.stuck"` constant added; `FORMATTERS[SESSION_STUCK]` renders tier, sinceMs (as seconds), lastTool, lastToolError
- `packages/daemon/src/claude-session/stuck-detector.ts`: New `StuckDetector` class — `recordProgress()` resets timer, tier escalation (90s/180s/300s → repeat), `performance.now()` monotonic clock, permission/non-active suspension, `dispose()` on all exit paths; constructor validates non-empty ascending thresholds, positive `repeatMs`, and per-element positive-finite values
- `packages/daemon/src/claude-session/session-state.ts:84`: `lastToolCall: {name, errorMessage?, at}` field; `extractLastToolCall()` / `extractLastToolCallRaw()` extract last `tool_use` block; `respondToPermission` sets `errorMessage` only when `request.tool_name === lastToolCall.name` (mismatched tools leave it untouched)
- `packages/daemon/src/claude-session/ws-server.ts`: `stuckDetector: StuckDetector | null` per `WsSession`; `ensureStuckDetector()`, `recordSessionProgress()`, `disposeStuckDetector()`, `handleStuckEvent()` methods; `disposeStuckDetector` called on `session:result`, `session:error`, `session:cleared`, `session:disconnected` (in `handleSessionEvent`), `terminateSession`, and `disconnectSessionWs`; `recordSessionProgress` called on `session:init`, `session:response`, `session:permission_request`, and for `tool_progress`/`stream_event` messages (prevents false positives during long-running tools/streaming); `handleStuckEvent` wrapped in try/catch; `SESSION_STUCK` forwarded to `onMonitorEvent` bus; `SessionWaitEvent` extended with `tier`, `sinceMs`, `tokenDelta`, `lastTool`, `lastToolError`

### Test Coverage
- `stuck-detector.spec.ts`: 14 unit tests — tier escalation 1→2→3→repeat, progress reset, permission suspension (`pendingPermissionCount > 0` and `state === "waiting_permission"`), non-active state stops detection, tokenDelta tracking between emissions, lastTool/lastToolError from snapshot, dispose idempotency, timer leak prevention, recordProgress-after-dispose no-op, flapping (no dup tier after flap), constructor validation (empty thresholds, non-ascending, non-positive per-element, non-positive repeatMs)
- `session-state.spec.ts` (`lastToolCall` describe block): 6 tests — starts null, extracts last tool_use block, takes last when multiple blocks, leaves unchanged with no tool_use, permission denial sets errorMessage (matching tool), mismatch tool denial does not set errorMessage
- `ws-server.spec.ts` (`stuck detector integration` describe block): 5 integration tests — stuck event fires after stall threshold (with tier/sessionId/sinceMs), no timer leak after `session:result`, no timer leak after bye (`session:ended`), workItemId propagation, `waitForEvent` resolves with full stuck diagnostic fields (tier/sinceMs/tokenDelta/lastTool/lastToolError)

### Test Results
- `check` job: typecheck ✓, lint ✓, test (non-daemon) ✓ (3906 pass, 22 todo, 0 fail), test (daemon) ✓ (2059 pass, 2 skip, 0 fail)

### CI Status
All 3 required checks pass on run [24911482608](https://github.com/theshadow27/mcp-cli/actions/runs/24911482608):
- `build` ✓ (13s)
- `check` ✓ (4m6s)
- `coverage` ✓ (6m0s)

Note: The initial coverage run failed with exit code 1 and truncated output (74KB of 480KB expected), consistent with a transient Bun crash tracked in [#1004](https://github.com/theshadow27/mcp-cli/issues/1004). Re-run passed cleanly.

### Copilot Gate
- Unreplied inline threads: **0** (all 6 prior threads addressed across 2 repair rounds)
- CHANGES_REQUESTED reviews: **none**

### Remaining Work (non-blocking)
- MCP execution errors (tool throws, non-zero exit) are architecturally invisible at the NDJSON layer and not surfaced via `lastToolError`; documented in `session-state.ts` with follow-up tracked per issue description
- Negative `Bun.sleep`-based assertions in tests use 4× threshold margins (acknowledged in prior repair delta as acceptable)

### Outcome
**qa:pass** — All described functionality implemented and verified. All acceptance criteria met. Tests green (14 unit + 6 session-state + 5 integration). CI fully green after transient Bun crash retry.
